File: src/main/java/org/example/kalkulationsprogramm/repository/AnfrageDokumentRepository.java
package org.example.kalkulationsprogramm.repository;

import org.example.kalkulationsprogramm.domain.AnfrageDokument;
import org.example.kalkulationsprogramm.domain.AnfrageGeschaeftsdokument;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface AnfrageDokumentRepository extends JpaRepository<AnfrageDokument, Long> {
    List<AnfrageDokument> findByAnfrageId(Long anfrageId);

    @Query("SELECT g FROM AnfrageGeschaeftsdokument g")
    List<AnfrageGeschaeftsdokument> findAllGeschaeftsdokumente();

    @Query("SELECT g FROM AnfrageGeschaeftsdokument g WHERE g.geschaeftsdokumentart = 'Rechnung'")
    List<AnfrageGeschaeftsdokument> findOffeneGeschaeftsdokumente();

    Optional<AnfrageDokument> findByGespeicherterDateiname(String gespeicherterDateiname);

    Optional<AnfrageDokument> findByGespeicherterDateinameIgnoreCase(String gespeicherterDateiname);

    /**
     * Liefert für eine Liste von Anfrage-IDs ein Mapping
     * [AnfrageGeschaeftsdokument.id, Anfrage.id], damit der DokumentFreigabeService
     * den jüngsten Freigabe-Status pro Anfrage finden kann.
     */
    @Query("SELECT g.id, g.anfrage.id FROM AnfrageGeschaeftsdokument g WHERE g.anfrage.id IN :anfrageIds")
    List<Object[]> findGeschaeftsdokumentIdMappingByAnfrageIds(@org.springframework.data.repository.query.Param("anfrageIds") List<Long> anfrageIds);
}

File: src/main/kotlin/org/example/kalkulationsprogramm/repository/AnfrageDokumentRepository.kt
package org.example.kalkulationsprogramm.repository

import java.util.Optional
import org.example.kalkulationsprogramm.domain.AnfrageDokument
import org.example.kalkulationsprogramm.domain.AnfrageGeschaeftsdokument
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface AnfrageDokumentRepository : JpaRepository<AnfrageDokument, Long> {
    fun findByAnfrageId(anfrageId: Long?): List<AnfrageDokument>

    @Query("SELECT g FROM AnfrageGeschaeftsdokument g")
    fun findAllGeschaeftsdokumente(): List<AnfrageGeschaeftsdokument>

    @Query("SELECT g FROM AnfrageGeschaeftsdokument g WHERE g.geschaeftsdokumentart = 'Rechnung'")
    fun findOffeneGeschaeftsdokumente(): List<AnfrageGeschaeftsdokument>

    fun findByGespeicherterDateiname(gespeicherterDateiname: String): Optional<AnfrageDokument>

    fun findByGespeicherterDateinameIgnoreCase(gespeicherterDateiname: String): Optional<AnfrageDokument>

    @Query("SELECT g.id, g.anfrage.id FROM AnfrageGeschaeftsdokument g WHERE g.anfrage.id IN :anfrageIds")
    fun findGeschaeftsdokumentIdMappingByAnfrageIds(
        @Param("anfrageIds") anfrageIds: List<Long>,
    ): List<Array<Any>>
}
